fix: make Statement.cancel and Statement.setQueryTimeout thread safe

vlsi opened the pull request:
Statement.cancel was not thread-safe, thus query timeout from one statement might cancel subsequent statement.

The change introduces state tracking of a statement (ide, in-query, cancelling, cancelled), so out-of-order cancels can be detected and ignored.

There are two basic problems with .cancel handing:
1) `cancel`s called at random from user code. Even thought `Statement#cancel` does not require thread-safety, it would be "a bit" unexpected if `statementA.cancel` would cancel subsequent `statementB` that is run in the same connection.
The thing is backend protocol does not allow to cancel a specific execution, thus if user invokes `cancel`, pgjdbc should ensure that cancel would not terminate another statement by mistake.

The approach is to have `statementState` and skip cancels that appear when state is "idle".
Additionally, if .execute notices there was a cancel, it waits till cancel completes.

Note: this dance does not guarantee that cancel is indeed dispatched to the backend, however the case when "cancel request" is not dispatched should be very rare. The remaining race is due to OS signals. In other words, current implementation just waits till the "interrupt" signal is sent to the proper backend, however it does not wait till the signal is processed.

2) `setQueryTimeout` should not affect another statements or subsequent executions of the same statement.

The approach here is to atomically `compareAndSwap` `cancelTimerTask` reference, so the task can detect presence of another task and exit safely.

vlsi commented:
I think PR is ready for review and merge.

davecramer commented:
Very elegant solution! Thanks.

Can others please read this over? I'd appreciate more eyes on this.

schlosna commented on org/postgresql/jdbc2/AbstractJdbc2Statement.java:
```diff
@@ -768,14 +821,14 @@ public int getQueryTimeout() throws SQLException
      * @param seconds - the new query timeout limit in seconds
```
Param is now millis, s/seconds/milliseconds/

vlsi replied:
good catch

vlsi commented:
Question: what is the expected behavior for multiple cancels in a row?
My current change lets only a single cancel per statement execution (subsequent cancels get ignored). Is it an expected behavior? Should multiple cancel requests be allowed?

davecramer commented:
@vlsi I think it's sufficient to just ignore subsequent cancel requests. The use case of cancel is for cancel to be called from a different thread than the one that the statement is currently executing on. I can't see a use case for multiple cancels to be accepted. Furthermore this would add considerable complexity

vlsi commented:
Small change: I've changed `synchronized(Statement)` -> `synchronized(Connection)` and added `synchronized(Connection)` for the duration of `.cancel`.
This makes `wait till cancel completes` burn less cpu.
